staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -22,11 +22,15 @@ import (
 	"k8s.io/apiserver/pkg/authentication/user"
 )
 
-// Token checks a string value against a backing authentication store and returns
-// information about the current user and true if successful, false if not successful,
-// or an error if the token could not be checked.
+// Token checks a token string and expected audiences against a backing
+// authentication store and returns information about the current user and true
+// if successful, false if not successful, or an error if the token could not
+// be checked. A token handler that is not audience aware, may choose to ignore
+// the audience argument. A nil or empty audience argument means that no
+// audiences were specified. In that case, an audience aware token
+// authenticator should reject the token.
```

Comment:
> A token handler that is not audience aware may choose to ignore the audience argument.

That would mean many of the existing token authenticators would return true for any audience. Ideally, those would only return true when we were checking token validity for the apiserver, not arbitrary clients.

> A nil or empty audience argument means that no audiences were specified.

Under what circumstances would we not pass audience in to the authenticator? The two entry points I know of are the authentication filter and the token review handler. I'd expect both of those places to specify at least one audience. When satisfying token review requests, aren't we taking empty audience to mean "valid against the API server"?

I was hoping we could get to something like this:

|                                                | auds=["","https://kubernetes.default.svc","my-apiserver-oidc-client"] | auds=["myapp"] |
|------------------------------------------------|-----------------------------------------------------------------------|----------------|
| old serviceaccount token                       | ok                                                                    |                |
| new serviceaccount token aud=https://kubernetes.default.svc         | ok                                                                    |                |
| new serviceaccount token aud=myapp             |                                                                       | ok             |
| bootstrap token                                | ok                                                                    |                |
| oidc with client-id=my-apiserver-oidc-client   | ok                                                                    |                |
| token-file token                               | ok                                                                    |                |
| webhook audience unaware                       | ok                                                                    |                |
| webhook audience aware, apiserver token, aud=? | ok (how do we know what to include in auds to make the remote endpoint pass?)                                                                    |                |
| webhook audience aware, myapp token            |                                                                       | ok             |

Notes:
* `auds` is always set when calling `AuthenticateToken`
* when authenticating tokens used against the apiserver, `""` could be included in the list of acceptable audiences, giving token authenticators that are audience-unaware a signal they can authenticate.
* the OIDC token authenticator is already audience-aware. we might need to include the `--oidc-client-id` parameter value in the set of audiences the apiserver identifies itself as
* it would be good to find a way to make an audience-unaware webhook authenticators only authenticate tokens for the apiserver, not arbitrary audiences. one possibility is having them echo the accepted audience in the TokenReview status, defaulting to "", and verified to be in the list of submitted audiences by the webhook authenticator. when checking for the apiserver, no audience would be echoed in the status, defaulting to "", which would be in the list of acceptable candidates. when checking for myapp, the defaulted accepted audience of "" in the status would not match, and would prevent the authentication from succeeding.
* similar to the OIDC authenticator, webhook authenticators may already be audience-aware, but we have no way of knowing what audience it might already be verifying inside the tokens sent to it. If we're using an audience-aware webhook authenticator, we need to think about how to include the audience it is verifying in our list of "audiences the apiserver identifies as"

Reply:
I think we need clarity over what gets passed for `auds` when:
* authenticating a token request to the API server
* handling a TokenReview with `Audiences:[]`
* handling a TokenReview with `Audiences:[foo]`

Reply:
it might be worthwhile to hoist this table to the design doc for the audience changes, to illustrate how we plan to keep compatibility with existing auth methods and APIs, while improving audience capabilities for webhook and service account authenticators